### PR TITLE
Add qs encoding options; Fix #2570

### LINKED ISSUE #2570: qs encoding works incorrectly
### Summary
I tried to send query parameters without encoding (i.e. single quote should be ```'``` instead of ```%27```). Used for this object-parameter ```qs``` with ```qsStringifyOptions = {encode: false}```, but the out request  came with encoding

image

Looked at the source code
request/request.js
```
1166: var qs = self._qs.stringify(base)

1172: self.uri = url.parse(self.uri.href.split('?')[0] + '?' + qs)
```

On 1166 line everything is OK, the qs returns without encoding:
```
title=AAAAAAAAAA'&qs1=BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB%20://ас'"д
```
But on 1172 line data is transmitted in the url.parse(), this method encoded all input data:
```
// it`s self.uri object
Url {
  protocol: 'https:',
  slashes: true,
  auth: null,
  host: 'ru.wikipedia.org:443',
  port: '443',
  hostname: 'ru.wikipedia.org',
  hash: null,
  search: '?title=AAAAAAAAAA%27&qs1=BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB%20://ас%27%22д',
  query: 'title=AAAAAAAAAA%27&qs1=BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB%20://ас%27%22д',
  pathname: '/w/index.php',
  path: '/w/index.php?title=AAAAAAAAAA%27&qs1=BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB%20://ас%27%22д',
  href: 'https://ru.wikipedia.org:443/w/index.php?title=AAAAAAAAAA%27&qs1=BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB%20://ас%27%22д' }
```

Am I doing something wrong or is it still a bug?

### Simplest Example to Reproduce

```js
request({
  method: 'GET',
  url: 'http://example.com', // a public URL that we can hit to reproduce, if possible
  qs: {
    qs1: "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB%20://ас'\"д"
  },
  qsStringifyOptions: {
    encode: false
  }
},
```

### Possible Solution
Need to check qsStringifyOptions

| software         | version
| ---------------- | -------
| request          | ^2.79.0
| node             | v6.10.0
| npm              | 3.10.10
| Operating System | Ubuntu 16.04 x64

### PR DESCRIPTION
**Pull request by wotgl**
## PR Checklist:
- [x] I have run `npm test` locally and all tests are passing.
- Fixes #2570

## PR Description
Added ```options``` parameter to ```Request.prototype.qs``` function. This ```options``` includes ```qsStringifyOptions``` from request options

### CODE REVIEW

File: request.js [hunk 1]
```diff
@@ -1244,7 +1248,14 @@ Request.prototype.qs = function (q, clobber) {
     return self
   }
 
-  self.uri = url.parse(self.uri.href.split('?')[0] + '?' + qs)
+  if (options && options.encode == false) {
```
Comment (Huachao) [inline request.js]: ===

**Comment by simov** [comment]
See https://github.com/request/request/issues/2570#issuecomment-288684825

**Comment by ahmdsolmn** [comment]
@simov https://github.com/request/request/issues/2570#issuecomment-288684825 doesn't solve the issue

**Comment by simov** [comment]
Ok, I see now. Add a test for it and I'll review the code again.

**Pull request closed without merging**